### PR TITLE
feat: Menu refactor to Fiori3

### PR DESCRIPTION
**Pull request by kavya-b**
## Related Issue
Closes SAP/fundamental-styles#571

## Description
Update Menu styles, documentation and tests to align with Fiori3 specs

- removed `fd-menu__group` classes, their playground tests and documentation examples related to group headers.
- added new classes for compact and mobile modes: `fd-menu--compact` and `fd-menu--mobile`. Default will be cozy tablet mode.
- moved `fd-menu__item--separated` from `<ul>` level to its own element `fd-menu__separator` in a `<span>` since separators are now used instead of group headers to show grouping.
- using existing class `fd-menu__title` for the text that will be used in a `<span>`
- added new class for  secondary icon `fd-menu__addon-after` that will be used similar to `fd-menu__addon-before`
- added a modifier class for the affordance arrow when used in a submenu `fd-menu__addon-after--submenu` to get the right(or left in RTL) navigation arrow in a smaller icon size. 
- added new class for submenu `fd-menu__sublist` at the `<ul>` level corresponding to `fd-menu__list` of the parent menu.
- added new class `fd-menu__link` that wraps the item name and associated icons, if any.
- added new class `has-child` that can be applied on the parent `__link` class to apply styles for the parent element that has a submenu.
- moved `fd-menu__item` class to the parent `<li>` tag.

## Breaking changes
- Removed `fd-menu__item--separated`. Instead use `fd-menu__separator` in its own `<span>` after the `<li>` item where you want the separator.
- Removed `fd-menu__group`. Separator itself acts a visual cue for grouping.
- Removed `fd-menu--addon-before`(not `fd-menu__addon-before`) as it is unnecessary now and will be covered by existing classes.
- Moved `fd-menu__item` class to the parent `<li>` tag. Use `fd-menu__link` class on the `<a>` tag instead.

For new classes added, please refer to the above section.

### Before:
image
image

### After:
image
image

### Updated screenshots:
image

image

### CODE REVIEW

File: src/menu.scss [hunk 1]
```diff
@@ -5,165 +5,303 @@
...
-    @include fd-weight("normal");
+    &--separated {
+      border-bottom-style: solid;
+      border-bottom-width: 0.0625rem;
+      border-bottom-color: $fd-menu-item-separator-color;
+    }
 
-    text-decoration: none;
+    @include fd-hover() {
+      .#{$block}__option-name {
```
Comment (JKMarkowski) [inline src/menu.scss]: Could you change this element name to `__title`?

File: src/menu.scss [hunk 2]
```diff
@@ -5,165 +5,303 @@
...
+  &__list,
+  &__sublist {
     @include fd-reset();
 
+    margin: 0;
+    padding: 0;
     list-style: none;
-    margin-bottom: 0;
-    padding-left: 0;
+    max-width: 20rem;
```
Comment (JKMarkowski) [inline src/menu.scss]: Let's put `max-width` to parent `fd-menu`  container

File: src/menu.scss [hunk 3]
```diff
@@ -5,165 +5,303 @@
...
+    }
+
+    @include fd-screen(l) {
+      width: 2rem; // to account for item right padding of 12px or 0.75rem so 2rem-0.75rem = 1.25rem
+      height: 1.25rem; // to account for item top padding of 6px or 0.375rem so 2rem-0.375rem-0.375rem = 1.25rem
+      margin: 0 0 0 -0.5rem;
+
+      @include fd-rtl() {
+        margin: 0 -0.75rem 0 0; // since before and after icons switch places, their margins also switch
+      }
```
Comment (JKMarkowski) [inline src/menu.scss]: Let's not use `fd-screen`, we let users to add size modes by themselves on `fd-menu` level. so ex. `fd-menu--compact`.

File: src/menu.scss [hunk 4]
```diff
@@ -5,165 +5,303 @@
...
+
+      @include fd-hover() {
+        color: $fd-menu-item-color--active;
+      }
+    }
+
+    flex: 0 0 auto;
+    display: flex;
     background-color: transparent;
     float: left;
```
Comment (JKMarkowski) [inline src/menu.scss]: Try to remove floats, it's out of control. There are better alternatives

File: src/menu.scss [hunk 5]
```diff
@@ -5,165 +5,303 @@
...
+    @include fd-selected() {
+      color: $fd-menu-primary-icon-color;
+
+      @include fd-hover() {
+        color: $fd-menu-item-color--active;
+      }
+    }
+
+    background-color: transparent;
+    float: right;
```
Comment (JKMarkowski) [inline src/menu.scss]: Again float, which is not needed

File: src/menu.scss [hunk 6]
```diff
@@ -5,165 +5,303 @@
...
+
+    @include fd-selected() {
+      color: $fd-menu-primary-icon-color;
+
+      @include fd-hover() {
+        color: $fd-menu-item-color--active;
+      }
+    }
+
+    flex: 0 0 auto;
```
Comment (JKMarkowski) [inline src/menu.scss]: It's also not needed I think.

File: src/menu.scss (hunk 3 above)
Comment (kavya-b) [reply src/menu.scss]: I was handling cozy, compact modes for tablet, desktop and mobile through this. If we provide option to user to add size modes, then we won't have control if they decide to use compact mode on phone thereby not aligning to Fiori3, right?

**Comment by kavya-b** [comment]
> ### The menu item has text decorator this doesn't match fiori 3
> image

**This is coming from the styles to <a> tag in documentation itself. It appears fine in playground:**
image
 
> ### wrong width in IE 11
> image

I will work on this; however, in the meanwhile, do you think this is related to https://github.com/SAP/fundamental-styles/pull/691 ?

> ### Could you add RTL example to documentation?

I will add this.

**Comment by kavya-b** [comment]
> * Let's add cozy mode by default, because right now there is only `compact` mode.

Do you mean, I should have the cozy mode applied by default? But for desktop, it is mentioned as Compact mode, right?
> * There are also differencted between Compact/Cozy(tablet)/Cozy(Phone)
>   image

Currently, the styles of cozy, compact are applied when the screen is resized. However, from your other comment, it seems that these styles are better provided as separate `fd-menu-cozy` `fd-menu-compact` classes. I will work on changing this.
> * Background color of `fd-menu` is not set

I assumed that since fd-menu will be part of a popover, the background color, box shadow and border attributes of the container should be part of the popover. I have verified that the appropriate values are applied to the popover class. 
> * Remove underline from `<a>` if used with `__title`
>   image

This is coming from the styles `<a>` tag in documentation itself. It appears fine in playground:
image
Currently same thing is happening for side-navigation component as well.

File: src/menu.scss (hunk 3 above)
Comment (JKMarkowski) [reply src/menu.scss]: Actually, in all of cases, when we have more than 1 `size mode` such as cozy/compact, we let users decide on application level, which one they want to use it.

Comment (JKMarkowski) [reply src/menu.scss]: Actually, in all of cases, when we have more than 1 `size mode` such as cozy/compact, we let users decide on application level, which one they want to use it.

**Comment by rengare** [comment]
> I will work on this; however, in the meanwhile, do you think this is related to #691 ?
Frankly, I don't know, we know that it is due to the popover and input used together but maybe it is related. I would have to check that

File: src/menu.scss (hunk 6 above)
Comment (kavya-b) [reply src/menu.scss]: removed

File: src/menu.scss (hunk 5 above)
Comment (kavya-b) [reply src/menu.scss]: removed

File: src/menu.scss (hunk 4 above)
Comment (kavya-b) [reply src/menu.scss]: removed

File: src/menu.scss (hunk 3 above)
Comment (kavya-b) [reply src/menu.scss]: added `fd-menu--compact-desktop` and `fd-menu--mobile` classes. Tablet cozy is default.

File: src/menu.scss [hunk 7]
```diff
@@ -5,165 +5,303 @@
...
-    @include fd-disabled() {
-      @include fd-var-color("outline-color", fd-color-state("disabled", "action"), --fd-color-action-disabled);
-      @include fd-var-color("color", $fd-forms-color--disabled, --fd-color-text-4);
+    @include fd-item-responsive-padding();
 
-      cursor: not-allowed;
+    .#{$block}__option-name {
+      @include fd-reset();
+
+      flex: 3 3 10%;
```
[the remark this answers is not in the document]
Comment (kavya-b) [reply src/menu.scss]: we need this to ensure the primary, secondary icons and title inside a menu item are properly placed. If removed, the secondary icon comes to the end of the title and misaligns.

File: src/menu.scss (hunk 2 above)
Comment (kavya-b) [reply src/menu.scss]: have moved to fd-menu

File: src/menu.scss (hunk 1 above)
Comment (kavya-b) [reply src/menu.scss]: done

**Comment by kavya-b** [comment]
There are 2 known issues in IE as of now that I couldn’t find a fix for but I also see the same issue is present in components that use List component(Select, Multi Input etc)-
1. focus border outline-offset property is not applied properly in IE due to which it sometimes does not appear properly
2. clicking on the parent <li>  area alone activates the states (active, selected etc). If you click on an area that is a child of <li> the states don’t show up.

**Comment by stefanoScalzo** [comment]
Screen Shot 2020-03-04 at 8 05 15 PM
disabled should not change color on hover and on click

**Comment by kavya-b** [comment]
> Screen Shot 2020-03-04 at 8 05 15 PM
> disabled should not change color on hover and on click

fixed

File: src/menu.scss [hunk 8]
```diff
@@ -5,165 +5,295 @@
...
   }
 
   &__item {
     @include fd-reset();
-    @include fd-ellipsis();
 
-    display: block;
+    display: flex;
+    align-items: center;
     transition: all $fd-animation-speed ease-in;
```
Comment (JKMarkowski) [inline src/menu.scss]: There is no mention about transition in menu specs

File: src/menu.scss [hunk 9]
```diff
@@ -5,165 +5,295 @@
...
+      @include fd-hover() {
+        color: $fd-menu-item-color--active;
+      }
+    }
+
+    display: flex;
+    background-color: transparent;
+
+    @include fd-icon-size("m", "before");
+    @include fd-icon-size("m", "after");
```
Comment (InnaAtanasova) [inline src/menu.scss]: This mixin makes the icon 16px. We should use `rem`
Screen Shot 2020-03-09 at 1 50 57 PM
Screen Shot 2020-03-09 at 1 50 49 PM

File: src/menu.scss [hunk 10]
```diff
@@ -5,165 +5,295 @@
...
     }
 
     background-color: transparent;
-    float: left;
-    clear: left;
-    min-height: 1px;
-    width: 30px;
+
+    @include fd-icon-size("s", "before");
+    @include fd-icon-size("s", "after");
```
Comment (InnaAtanasova) [inline src/menu.scss]: This size should be applied only if the user wants affordance arrow. It shouldn't be applied for all addon-after. Maybe create a modifier class to handle the affordance arrow case

Comment (kavya-b) [reply src/menu.scss]: so, would you say `fd-menu__addon-after` by default takes any icon in the normal font-size of 1rem but for affordance arrow I should keep a class like `fd-menu__addon-after--submenu` and handle the smaller size case here?

File: src/menu.scss (hunk 9 above)
Comment (kavya-b) [reply src/menu.scss]: Okay, I will just use `var(--sapFontLargeSize)` directly instead

File: src/menu.scss (hunk 8 above)
Comment (kavya-b) [reply src/menu.scss]: Missed this. Will remove it, thanks for pointing it out!

**Comment by kavya-b** [comment]
> Screen Shot 2020-03-09 at 2 01 41 PM
> 
> I don't think it's a good approach to add the separator to the item. It should be a separate element. In the case where we have an item that is a separator and is selected, the selected border overwrites the separator. It should be like this: (has both lines)
> Screen Shot 2020-03-09 at 2 01 28 PM

I see. I will add a separate element independent of the item that is placed after the item.
> 
> * `fd-menu--compact-desktop` should be `fd-menu--compact` for consistency with the other components

Will change this
> * In Mobile mode how do we display the cascading (the second menu level) ?
> 
> Screen Shot 2020-03-09 at 2 05 50 PM
> 
The spec says _'Both controls are cascading, except for the sap.m.Menu on phone, where it opens fullscreen (with drill-drown navigation).'_ So the submenu will look the same way as the parent menu. 
> * For the documentation part - would you please remove the `Affordance arrow` if there is no second level. It gives the impression that something is not working because clicking on it does not open anything.

I will remove the arrows and instead show other icons to show how secondary icons look.

**Comment by kavya-b** [comment]
Addressed @JKMarkowski and @InnaAtanasova's comments: 
- removed redundant right-arrow icons when submenu was not really present. For submenu, use `fd-menu__addon-after--submenu` class modifier to get the smaller size for affordance arrow:
`<span class="fd-menu__addon-after fd-menu__addon-after--submenu sap-icon--navigation-right-arrow"></span>`
- instead used other icons to showcase secondary icon size
image
- using separator in a different element
image
- removed 'desktop' from modifier class name
- removed transition property
- applied font-size on icon directly
- updated playground tests and screenshots accordingly - added selected-separated usecase to visual test and documentation
image

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-03-11 at 11 08 11 AM

Hi @kavya-b , 
`fd-menu__item--separated` is contradicting the BEM methodology. The separator is not an menu item, which means we can not use `fd-menu__item`. Since it's not a menu item and is not a modifier of this item we can not use the `--separated`. You can rename it to something like `fd-menu__separator`.
For more info about BEM you can check [this](http://getbem.com/introduction/)

**Comment by kavya-b** [comment]
> Screen Shot 2020-03-11 at 11 08 11 AM
> 
> Hi @kavya-b ,
> `fd-menu__item--separated` is contradicting the BEM methodology. The separator is not a menu item, which means we can not use `fd-menu__item`. Since it's not a menu item and is not a modifier of this item we can not use the `--separated`. You can rename it to something like `fd-menu__separator`.
> For more info about BEM you can check [this](http://getbem.com/introduction/)

Okay, have changed the name to `fd-menu__separator` now.

**Comment by droshev** [comment]
@kavya-b Few comments:
- can you update the documentation and create examples without edge cases like long text items; hybrid items of text and icons - those can be added as part of the playground so can have testing examples there. The goal of the documentation is to show correct and recommended layouts without too long text or inconsistent icon/texts
- would you mind documenting the class names that the menu supports (e.g. `fd-menu__addon-before`, `fd-menu__addon-after`, `fd-menu__separator`,  all state classes, `fd-menu__sublist`)

File: src/menu.scss [hunk 11]
```diff
@@ -5,165 +5,294 @@
...
+
+      background-color: $fd-menu-item-background-color--active;
+
+      .#{$block}__addon-before,
+      .#{$block}__addon-after {
+        color: $fd-menu-primary-icon-color-active-selected;
+      }
+
+      .#{$block}__addon-after {
+        // border-bottom-color: var(--sapList_BorderColor); // TODO change this when fade() evaluates to a value
```
Comment (droshev) [inline src/menu.scss]: you can remove this comment

**Comment by kavya-b** [comment]
> @kavya-b Few comments:
> 
> * can you update the documentation and create examples without edge cases like long text items; hybrid items of text and icons - those can be added as part of the playground so can have testing examples there. The goal of the documentation is to show correct and recommended layouts without too long text or inconsistent icon/texts

okay, so for documentation, instead of showing every combination of single icon/double icon/long text, I keep only one format and use the same even for mobile mode and rtl mode? Should I keep the example with different states(active, hover, selected etc) or remove it as well?

> * would you mind documenting the class names that the menu supports (e.g. `fd-menu__addon-before`, `fd-menu__addon-after`, `fd-menu__separator`,  all state classes, `fd-menu__sublist`) - you can get inspired from here: https://fundamental-styles.netlify.com/components/bar.html

had added some of these in the PR description, but will update to the documentation description as well.
> * in the PR add a section `BREAKING CHANGES` and add all class names that were removed

Okay, will do this.

File: src/menu.scss (hunk 11 above)
Comment (kavya-b) [reply src/menu.scss]: done. have pushed in the other changes as well, please check.

**Comment by droshev** [comment]
@kavya-b thanks for the updates.
Could you try to review your component following the points from https://github.com/SAP/fundamental-styles/wiki/Self-Contained-Styles. There is something wrong with the separator:
Screen Shot 2020-03-18 at 2 31 03 PM

**Comment by kavya-b** [comment]
> https://github.com/SAP/fundamental-styles/wiki/Self-Contained-Styles

Hi @droshev, I tried to figure out if I am missing something that makes the unnormalized mode look like that, but cannot point to anything.
`fd-menu__separator` is in its own element in order to cover the edge case where we have a selected item and also need to show the separator, as pointed by @InnaAtanasova  in this comment: https://github.com/SAP/fundamental-styles/pull/702#pullrequestreview-371413725

**Comment by droshev** [comment]
@kavya-b We need to check how the nested menu behaves and looks on mobile. I guess it may behave like the nested list in side navigation. Can you follow up on that?
Meanwhile i fixed the self-contained issue I had raised before.

**Comment by kavya-b** [comment]
> @kavya-b We need to check how the nested menu behaves and looks on mobile. I guess it may behave like the nested list in side navigation. Can you follow up on that?

Hi @droshev , the [spec](https://wiki.wdf.sap.corp/wiki/pages/viewpage.action?pageId=2061501144#Menu(Fiori3)-MenuInteractionStates) says "_On Phone the Submenu opens in a new page, leaving the Menu option behind._ " The same visual design will apply for the submenu; on clicking of the parent item, the submenu will open in the same dialog, replacing the parent menu options, and the device's back button will bring it back to the parent menu. 
Does this need to be added to the documentation? I am not sure how to show a full screen replacement of parent menu in the same example using only aria-controls. I have something like this now: 
image
Please let me know if this suffices, I will push in the changes.
> Meanwhile i fixed the self-contained issue I had raised before.

Thank you for this.

**Comment by kavya-b** [comment]
@droshev As discussed, the documentation for mobile mode has been modified to reflect full screen:
submenu_mobile

**Pull request merged**